### ISSUE TITLE
getInstanceName omits the index when called from an instance of an array of models

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#11600]
Consider the following MWE:
```modelica
package TestGetInstanceName
  model M
    Real x = 1;
    String instName = getInstanceName();
  end M;
  
  model Test
    M m1;
    M m2;
    M m_1D[2];
    M m_2D[1,2];
  equation
    when initial() then
      Modelica.Utilities.Streams.print(m1.instName);
      Modelica.Utilities.Streams.print(m2.instName);
      Modelica.Utilities.Streams.print(m_1D[1].instName);
      Modelica.Utilities.Streams.print(m_1D[2].instName);
      Modelica.Utilities.Streams.print(m_2D[1, 1].instName);
      Modelica.Utilities.Streams.print(m_2D[1, 2].instName);
    end when;
  end Test;
end TestGetInstanceName;
```
When running Test, the following output is produced:
```
Test.m1 
Test.m2 
Test.m_1D 
Test.m_1D 
Test.m_2D 
Test.m_2D 
```
This output should be produced instead:
```
Test.m1 
Test.m2 
Test.m_1D[1] 
Test.m_1D[2]
Test.m_2D[1, 1] 
Test.m_2D[1, 2] 
```
This issue affects `Modelica.Blocks.Noise` models, when instantiating arrays of them, see #10475.

@perost could you please have a look?

**Comment by perost** [comment]
Fixed by  #11602, `getInstanceName` is now evaluated during the flattening instead of during the typing in order to have the subscripts included.

It should be noted though that [3.7.4.6](https://specification.modelica.org/master/operators-and-expressions.html#getinstancename) about `getInstanceName` says: "The simulation result should not depend on the return value of this function."

**Comment by casella** [comment]
Thanks @perost!

> It should be noted though that [3.7.4.6](https://specification.modelica.org/master/operators-and-expressions.html#getinstancename) about `getInstanceName` says: "The simulation result should not depend on the return value of this function."

I know, I know, I also thought about that while opening this ticket. Some people in MAP-Lang objected to the introduction of `getInstanceName()` for a long time because they thought that once this function was available, people might "abuse" it, whatever that meant. Eventually we got this function, with this safeguard clause against all kind of abuse. I never thought this was a problem, and I believe that using `getInstanceName()` to properly handle the seeds of random number generators is actually a clever idea and works just fine.

I'll open a ticket on the Modelica Specification arguing for the removal of that sentence from the specification.
